### DISCUSSION TITLE
Can I set  physical layer channel paraments,like SNR?

### DISCUSSION THREAD

**Discussion by wmwwimor** [srsran/srsRAN_Project#173] [General Help]
I use Zeromq to connect UE and gnb ,which is just a way to build a tcp tunnel ,right? 
I want to change sth in physical layer ,like setting  parameters about physical channel, is this  feasible?
Can I add some blocks by gnuradio which supports the deploy of zmq?
Or,,,I must use usrp  ?

**Comment by frankist** [comment]
Yes, you can alter the physical channel conditions using zmq + gnuradio. Here is an example of a script that creates an SNR staircase in gnuradio. You need to run first the gnb, then the script, then the ue, then press enter in the script, and you should see the SNR changing.

[snr_ramp_test.py.tar.gz](https://github.com/srsran/srsRAN_Project/files/12383393/snr_ramp_test.py.tar.gz)

**Reply by wmwwimor** [reply]
Actually, I want to make some changes in physical layer, like optimizing the LDPC coding,so any other physical channel parameters can I revise after modulation .I can only simulate the channel between UE and gnb by Gnuradio？How about config files in srsRAN source code ?
Thx~~

**Reply by wmwwimor** [reply]
I mean that I want to change the parameters of  physical layer channel of the whole system , not just the physical channel between UE and gnb.

**Reply by frankist** [reply]
You have to be more specific on what you want to change. Is it the mcs? If so, you can set it in the config file.

**Reply by wmwwimor** [reply]
The layer I'm working on is the higher physical, as you know ,after modulation ,the signal will be sent to physical communication channel ,and my aim is to optimize the packet loss in physical  layer by channel coding like LDPC ,before modulation ,and before channel coding the bit should be encoded by sourse coding. So can I change the coding way in srsRAN?of course ,I also want to simulate the performance of coding by change the communication channel parameter ,can I ?

**Reply by dvdgrgrtt** [reply]
At the moment, there aren't a lot of PHY algorithms that can be tweaked from the conf file. In the specific case of the LDPC decoder, you can decide the max number of decoding iterations and whether to use early stopping or not (after each iteration, the decoder checks the CRC and, if OK, stops without running all the iterations). The corresponding options are documented [here](https://docs.srsran.com/projects/project/en/latest/user_manuals/source/config_ref.html#expert-phy).

>So can I change the coding way in srsRAN?

As @frankist said, you can pick the MCS (see the _pdsch_ and _pusch_ sections of the [cell configuration](https://docs.srsran.com/projects/project/en/latest/user_manuals/source/config_ref.html#cell-cfg)). Beyond that (if you want to try alternative implementations of the encoder/decoder or, pushing even further, if you are working on a coding scheme different from the one in the standard), you'll have to modify the code.

**Reply by dvdgrgrtt** [reply]
By the way, if your focus is on the PHY layer and you have access to a MATLAB license, you can also have a look at our MATLAB toolbox, [srsRAN-matlab](https://github.com/srsran/srsRAN_matlab/). For instance, the [PUSCHBLER app](https://github.com/srsran/srsRAN_matlab/blob/main/apps/simulators/PUSCHBLER/PUSCHBLER.m) allows you to run link-level simulations of the PUSCH and estimate BLER and throughput using the srsRAN PUSCH decoder. The app already includes a channel model and its parameters are easily configurable.

**Comment by Zplusless** [comment]
Hi，Is it correct that both the zmq mode and the UHD mode have a complete gNB protocol stack, and the only difference lies in the method of wireless signal transmission (TCP sockets or over the air)?

**Reply by pgawlowicz** [reply]
yes

**Reply by Zplusless** [reply]
Thanks for your quick reply. According to [srsRAN gNB with srsUE — srsRAN Project documentation](https://docs.srsran.com/projects/project/en/latest/tutorials/source/srsUE/source/index.html), srsUE has the 15 kHz SCS and 10MHz bandwidth limitation, which limits the bitrates to around 30Mbps. Is there any solution or any other zmq-based based UE that can bypass this limitation? Thanks again!

**Reply by pgawlowicz** [reply]
yes, we have fixed srsUE to work with 5/10/15/20MHz (but always with 15kHz), the fix will be available with the next srsRAN_4g release.

**Reply by Zplusless** [reply]
Thanks! REALLY REALLY looking forward to the new release!

**Reply by Zplusless** [reply]
I have tried the 20MHz config but it failes to work. More details are here: https://github.com/srsran/srsRAN_Project/discussions/350#discussion-5921304
